testament/testament.nim

```diff
@@ -637,156 +658,194 @@ proc loadSkipFrom(name: string): seq[string] =
+  of "all": # Run all tests
+    var cats: seq[string]
+    var cmds: seq[string]
+    ## def qol procedure
+    proc progressStatus(idx: int) =
+      msg Progress: `%`("progress[all]: $1/$2 starting: cat: $3", [$idx, $cats.len, cats[idx]])
+    # Prepare myself
     var myself = quoteShell(getAppFilename())
-    if targetsStr.len > 0:
-      myself &= " " & quoteShell("--targets:" & targetsStr)
-
-    myself &= " " & quoteShell("--nim:" & compilerPrefix)
-    if testamentData0.batchArg.len > 0:
-      myself &= " --batch:" & testamentData0.batchArg
+    block prepare_myself:
+      if optFailing:
+        myself &= " " & quoteShell("--failing")
+      if retryContainer.retry:
+        myself &= " " & quoteShell("--retry")
+      if targetsStr.len > 0:
+        myself &= " " & quoteShell("--targets:" & targetsStr)
+      myself &= " " & quoteShell("--nim:" & compilerPrefix)
+      if testamentData0.batchArg.len > 0:
+        myself &= " --batch:" & testamentData0.batchArg
+      if skipFrom.len > 0:
+        myself &= " " & quoteShell("--skipFrom:" & skipFrom)
 
-    if skipFrom.len > 0:
-      myself &= " " & quoteShell("--skipFrom:" & skipFrom)
-
-    var cats: seq[string]
     let rest = if p.cmdLineRest.len > 0: " " & p.cmdLineRest else: ""
+    # Traverse the test directory
     for kind, dir in walkDir(testsDir):
       assert testsDir.startsWith(testsDir)
+      # The category name is extracted from the directory
+      # eg: 'tests/compiler' -> 'compiler'
       let cat = dir[testsDir.len .. ^1]
+      if retryContainer.retry and cat notin retryContainer.cats:
+        continue
       if kind == pcDir and cat notin ["testdata", "nimcache"]:
         cats.add cat
     if isNimRepoTests():
       cats.add AdditionalCategories
+    # User may pass an option to skip the megatest category, default is useMegaTest
     if useMegatest: cats.add MegaTestCat
+    # We now prepare the command line arguments for our child processes
 
-    var cmds: seq[string]
     for cat in cats:
+      # Remember that if we are performing the megatest category, then
+      # all joinable tests will be covered in that, so we use the parallel cat
+      # action
       let runtype = if useMegatest: " pcat " else: " cat "
       cmds.add(myself & runtype & quoteShell(cat) & rest)
-
-    proc progressStatus(idx: int) =
-      echo "progress[all]: $1/$2 starting: cat: $3" % [$idx, $cats.len, cats[idx]]
-
-    if simulate:
+      
+    if simulate: # 'see what tests would be run but don't run them (for debugging)'
       skips = loadSkipFrom(skipFrom)
       for i, cati in cats:
         progressStatus(i)
         processCategory(r, Category(cati), p.cmdLineRest, testsDir, runJoinableTests = false)
     else:
       addExitProc azure.finalize
-      quit osproc.execProcesses(cmds, {poEchoCmd, poStdErrToStdOut, poUsePath, poParentStreams}, beforeRunEvent = progressStatus)
-  of "c", "cat", "category":
+      let processOpts =
+        if optFailing and not optVerbose:
```

Comment:
I'm not sure how `optFailing` can override `optVerbose`?

Reply:
optVerbose implies more information. optFailing hides everything to not do with progress and failures; however, combined with optVerbose, you will still have information printed such as the command that was passed to the process to run that test (as per Saems request).